util/ibex_config.py

```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env python3
+
+import argparse
+import collections
+import os
+import sys
+
+import hjson
+
+DEFAULT_CONFIG_FILE = 'ibex_configs.hjson'
+
+
+class ConfigException(Exception):
+    pass
+
+
+def verify_config(name, config_dict):
+    """Check config_dict matches expectations:
+        - It's a mapping object e.g. OrderedDict
+        - It's values are either strings or integers"""
+    if not isinstance(config_dict, collections.Mapping):
+        raise ConfigException('Config ' + name +
+                              ' must have dictionary giving parameters')
+
+    for k, v in config_dict.items():
+        if isinstance(v, int):
+            continue
+        if isinstance(v, str):
+            continue
+
+        raise ConfigException('Parameter ' + k + ' for config ' + name +
+                              ' must be string or int got ' + str(type(v)))
+
+
+def get_config_dicts(config_file):
+    """From a open file config_file extract a dictionary of configuration
+    dictionaries
+
+    Throws ConfigException on any error"""
+    try:
+        config_hjson = hjson.load(config_file)
+
+        for k, v in config_hjson.items():
+            verify_config(k, v)
+
+        verify_config_parameters(config_hjson)
+
+        return config_hjson
+    except hjson.HjsonDecodeError as e:
+        raise ConfigException('Could not decode hjson ' + str(e))
+
+
+def verify_config_parameters(config_dicts):
+    """Verify parameters across config_dicts match expectations:
+        - Each config has the same set of parameters specified"""
+    parameters = set()
+
+    first = True
+
+    for name, config_dict in config_dicts.items():
+        parameters_this_config = set()
+
+        for parameter, value in config_dict.items():
+            if first:
+                parameters.add(parameter)
+
+            parameters_this_config.add(parameter)
+
+        if first:
+            first = False
+        else:
+            parameter_difference = parameters ^ parameters_this_config
+            if len(parameter_difference) != 0:
+                raise ConfigException('Config ' + name +
+                                      ' has differing parameters ' +
+                                      ','.join(parameter_difference))
+
+
+def config_dict_to_fusesoc_opts(config_dict):
+    fusesoc_cmd = []
+    for parameter, value in config_dict.items():
+        fusesoc_cmd.append('--' + parameter + '=' + str(value))
+
+    return ' '.join(fusesoc_cmd)
```

Comment:
For bonus points, you probably want to use `shlex.quote` here, assuming that the output is supposed to be something that the shell can run.

Reply:
Well it's something that should be appended to a fusesoc command line rather than run directly but shlex.quote should do something useful all the same, I'll add it in.